### PR TITLE
AppNexus: Render Ads Above the Fold

### PR DESCRIPTION
**Pull request by protonate**
render ads that are available but not rendered when the slot is above the fold and then scrolled into view

### CODE REVIEW

File: ads/appnexus.js [hunk 1]
```diff
@@ -108,14 +108,38 @@ function appnexusAst(global, data) {
...
+    && tagKeys.map(key => {
+      const tag = tagsRequested[key];
+      return !tag.showTagCalled && tag.adResponse;
+    }).filter(response => response);
+    if (responses && responses.length) {
+          responses.map(response => {
+        if (response.nobid) {
+          context.noContentAvailable();
+        } else {
+          const adObj = response && apntag.getAdForAMP(response);
```
Comment (mkendall07) [inline ads/appnexus.js]: is `getAdForAMP` a new function?

File: ads/appnexus.js [hunk 2]
```diff
@@ -108,14 +108,38 @@ function appnexusAst(global, data) {
     global.apntag = context.master.apntag;
   }
 
-  apntag.anq.push(() => {
-    apntag.onEvent('adAvailable', data.target, adObj => {
-      global.context.renderStart({width: adObj.width, height: adObj.height});
-      apntag.showTag(data.target, global.window);
+  // check for responses received before listeners are registered,
+  //  for example when an above-the-fold ad is scrolled into view
+  const tagsRequested = apntag.requests && apntag.requests.tags;
```
Comment (mkendall07) [inline ads/appnexus.js]: eventually `apntag.requests.tags` will go away, is there a way to make this a more generic public API?

**Review by mkendall07 — CHANGES_REQUESTED** [review]

File: ads/appnexus.js (hunk 1 above)
Comment (protonate) [reply ads/appnexus.js]: yep

File: ads/appnexus.js [hunk 3]
```diff
@@ -108,14 +108,24 @@ function appnexusAst(global, data) {
...
-    apntag.onEvent('adNoBid', data.target, () => {
-      context.noContentAvailable();
-    });
+  apntag.anq.push(() => {
+    apntag.onEvent('adAvailable', data.target, isAdAvailable);
+    apntag.onEvent('adNoBid', data.target, noAdAvailable);
   });
 }
+
+function isAdAvailable(adObj) {
```
Comment (mkendall07) [inline ads/appnexus.js:124]: shouldn't this be `adIsAvailable`?

Comment (protonate) [reply ads/appnexus.js:124]: this is consistent with conventions

**Pull request merged**